Mathlib/Analysis/NormedSpace/TrivSqZeroExt.lean

```diff
@@ -188,6 +172,94 @@ theorem eq_smul_exp_of_ne_zero (x : tsze R M) (hx : x.fst ≠ 0) :
   eq_smul_exp_of_invertible _ _
 #align triv_sq_zero_ext.eq_smul_exp_of_ne_zero TrivSqZeroExt.eq_smul_exp_of_ne_zero
 
-end NormedField
+end Field
+
+end Topology
+
+/-!
+### The $ℓ^1$ norm on the trivial square zero extension
+
+This is not particularly canonical, but is sufficient to obtain a normed algebra, and thus
+to use `exp_add_of_commute`.
+-/
+
+noncomputable section Normed
+
+section Ring
+
+variable [NormedField 𝕜] [NormedRing R] [NormedAddCommGroup M]
+variable [NormedAlgebra 𝕜 R] [NormedSpace 𝕜 M] [Module R M] [Module Rᵐᵒᵖ M]
+variable [BoundedSMul R M] [BoundedSMul Rᵐᵒᵖ M] [SMulCommClass R Rᵐᵒᵖ M]
+variable [IsScalarTower 𝕜 R M] [IsScalarTower 𝕜 Rᵐᵒᵖ M]
+
+instance instL1NormedAddCommGroup : NormedAddCommGroup (tsze R M) :=
+  inferInstanceAs <| NormedAddCommGroup (WithLp 1 <| R × M)
+
+theorem norm_def (x : tsze R M) : ‖x‖ = ‖fst x‖ + ‖snd x‖ := by
+  rw [WithLp.prod_norm_eq_add (by norm_num)]
+  simp only [ENNReal.one_toReal, Real.rpow_one, div_one]
+  rfl
+
+theorem nnnorm_def (x : tsze R M) : ‖x‖₊ = ‖fst x‖₊ + ‖snd x‖₊ := by
+  ext; simp [norm_def]
+
+instance instL1NormedSpace : NormedSpace 𝕜 (tsze R M) :=
+  inferInstanceAs <| NormedSpace 𝕜 (WithLp 1 <| R × M)
+
+instance instL1NormedRing : NormedRing (tsze R M) where
+  norm_mul
+  | ⟨r₁, m₁⟩, ⟨r₂, m₂⟩ => by
+    dsimp
+    rw [norm_def, norm_def, norm_def, add_mul, mul_add, mul_add, snd_mul, fst_mul]
+    dsimp [fst, snd]
+    rw [add_assoc]
+    gcongr
+    · exact norm_mul_le _ _
+    refine (norm_add_le _ _).trans ?_
+    gcongr
+    · exact norm_smul_le _ _
+    refine (_root_.norm_smul_le _ _).trans ?_
+    rw [← add_zero (_ * _), mul_comm, MulOpposite.norm_op]
+    gcongr
```

Comment:
It would be very helpful if `gcongr` knew how to apply `norm_smul_le` and `norm_add_le` here.

Reply:
If I'm not mistaken, you're pining for `grw` for which there's an open PR: #8167